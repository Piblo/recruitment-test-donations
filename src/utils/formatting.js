import { timeFormat } from 'd3';

export function formatDate(date) {
  const d = date instanceof Date
    ? date
    : new Date(date);

  return timeFormat('%d/%m/%Y')(d);
}
