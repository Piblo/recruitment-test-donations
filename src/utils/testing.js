import React from 'react';
import renderer from 'react-test-renderer';
import { ThemeProvider } from 'styled-components';
import { theme } from '../components';

export function renderWithTheme(component) {
  return renderer.create(
    <ThemeProvider theme={theme}>
      {component}
    </ThemeProvider>
  );
}
