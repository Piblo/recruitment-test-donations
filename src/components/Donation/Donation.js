import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Container, Avatar, Paper } from '..';
import { getCurrencySymbol, formatDate } from '../../utils/formatting';

const Donation = ({ donorName, amount, date, avatarUrl, currencyCode, message }) => {
  const [hovered, setHovered] = useState(false);

  const handleMouseOver = () => {
    setHovered(true);
  };

  const handleMouseLeave = () => {
    setHovered(false);
  };

  return (
    <StyledContainer alignItems="center" grow={1} onMouseOver={handleMouseOver} onMouseLeave={handleMouseLeave}>
      <Avatar src={avatarUrl} />
      <DonorName>{donorName}</DonorName>
      <Amount>{getCurrencySymbol(currencyCode) + amount}</Amount>
      <VerticalLine />
      <EventCircle />
      <DonationDate>{formatDate(date)}</DonationDate>
      {hovered && <Message>{message}</Message>}
    </StyledContainer>
  );
};

export default Donation;

Donation.propTypes = {
  donorName: PropTypes.string,
  amount: PropTypes.number,
  date: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
  avatarUrl: PropTypes.string,
  currencyCode: PropTypes.string,
  message: PropTypes.string
};

const StyledContainer = styled(Container)`
  position: relative;
`;

const DonorName = styled.p`
  margin-bottom: 6px;
`;

const Amount = styled.span`
  color: ${props => props.theme.palette.primary};
  font-size: 1.3em;
  margin-bottom: 6px;
`;

const DonationDate = styled.div`
  color: ${props => props.theme.palette.textLight};
  position: absolute;
  bottom: -32px;
  font-size: 0.8em;
`;

const VerticalLine = styled.div`
  border-left: 1px dashed ${props => props.theme.palette.secondary};
  height: 40px;
`;

const EventCircle = styled.div`
  border-radius: 50%;
  height: 15px;
  width: 15px;
  background-color: ${props => props.theme.palette.primary};
  transform: translateY(50%);
`;

const Message = styled(Paper)`
  position: absolute;
  left: 20px;
  transform: translateX(50%);
  width: 100%;
  z-index: 1;

  &:before {
    content: "";
    width: 0;
    height: 0;
    border-top: 10px solid transparent;
    border-bottom: 10px solid transparent;
    border-right:10px solid #4C4F60;
    position: absolute;
    left: -10px;
    top: 50%;
    transform: translateY(-50%);
  }
`;
