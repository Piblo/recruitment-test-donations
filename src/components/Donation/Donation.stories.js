import React from 'react';
import { storiesOf } from '@storybook/react';
import Donation from './Donation';

storiesOf('Donation', module)
  .add('default', () => (
    <Donation
      avatarUrl="https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif"
      donorName="Michael Gibson"
      currencyCode="GBP"
      date={new Date('2019-04-23')}
      amount={10}
      message="Every little counts!"
    />
  ));
