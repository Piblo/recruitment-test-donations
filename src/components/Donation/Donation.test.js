import React from 'react';
import { shallow } from 'enzyme';
import Donation from './Donation';
import { Avatar } from '..';
import { renderWithTheme } from '../../utils/testing';

describe('<Donation>', () => {
  it('Displays donor avatar', () => {
    const avatarUrl = 'https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif';

    const component = shallow(<Donation avatarUrl={avatarUrl} />);

    const avatarElement = component.find(Avatar);
    expect(avatarElement.prop('src')).toEqual(avatarUrl);
  });

  it('Displays donor name', () => {
    const donorName = 'Michae Gibson';
    const component = shallow(<Donation donorName={donorName} />);
    expect(component.contains(donorName)).toBeTruthy();
  });

  it('Displays donation amount', () => {
    const amount = 10;
    const currencyCode = 'GBP';
    const component = shallow(<Donation amount={amount} currencyCode={currencyCode} />);
    expect(component.contains(`Â£${amount}`)).toBeTruthy();
  });

  it('Displays donation date', () => {
    const date = new Date('2019-04-23');
    const component = shallow(<Donation date={date} />);
    expect(component.contains('23/04/2019')).toBeTruthy();
  });

  it('Matches snapshot', () => {
    const tree = renderWithTheme(
      <Donation
        avatarUrl="https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif"
        donorName="Michael Gibson"
        currencyCode="GBP"
        date={new Date('2019-04-23')}
        amount={10}
      />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
