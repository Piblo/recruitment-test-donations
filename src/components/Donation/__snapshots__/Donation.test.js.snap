// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Donation> Matches snapshot 1`] = `
<div
  className="Container-sc-13tjqbm-0 StyledContainer-sc-12yhrow-0 bitGVB"
  direction="column"
  onMouseLeave={[Function]}
  onMouseOver={[Function]}
>
  <img
    className="Avatar-sc-1o0a796-0 fZQRzI"
    src="https://www.justgiving.com/content/images/graphics/icons/avatars/facebook-avatar.gif"
  />
  <p
    className="DonorName-sc-12yhrow-1 gXLUOY"
  >
    Michael Gibson
  </p>
  <span
    className="Amount-sc-12yhrow-2 jFAgTE"
  >
    Â£10
  </span>
  <div
    className="VerticalLine-sc-12yhrow-4 ellmqY"
  />
  <div
    className="EventCircle-sc-12yhrow-5 eSQixU"
  />
  <div
    className="DonationDate-sc-12yhrow-3 efGwlT"
  >
    23/04/2019
  </div>
</div>
`;
